bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptorTest.java

```diff
@@ -0,0 +1,195 @@
+      RefreshingOAuth2CredentialsInterceptor.getCacheState(underTest.headerCache.get()));
+    long expiredStaleDiff =
+        HeaderCacheElement.TOKEN_STALENESS_MS - HeaderCacheElement.TOKEN_EXPIRES_MS;
+    setTime(2L + expiredStaleDiff);
+    Assert.assertEquals(CacheState.Expired,
+      RefreshingOAuth2CredentialsInterceptor.getCacheState(underTest.headerCache.get()));
+  }
+
+  @Test
+  /**
+   * Test that checks that concurrent requests to RefreshingOAuth2CredentialsInterceptor refresh
+   * logic doesn't cause hanging behavior.  Specifically, when a 
+   */
+  public void testRefreshDoesntHang() throws Exception,
+      TimeoutException {
+    // Assume that the user starts at this time... it's an arbitrarily big number which will
+    // assure that subtracting HeaderCacheElement.TOKEN_STALENESS_MS and TOKEN_EXPIRES_MS will not 
+    // be negative.
+    long start = HeaderCacheElement.TOKEN_STALENESS_MS * 10;
+    setTime(start);
+    
+    // RefreshingOAuth2CredentialsInterceptor will show that the access token is stale.
+    final long expiration = start + HeaderCacheElement.TOKEN_EXPIRES_MS + 1;
+    
+    // Create a machanism that will allow us to control when the accessToken is returned.  
+    // credentials.refreshAccessToken() will get called asynchronously and will wait until the 
+    // lock is notified before returning.  That will allow us to set up multiple concurrent calls
+    final Object lock = new String("");
+    Mockito.when(credentials.refreshAccessToken()).thenAnswer(new Answer<AccessToken>() {
+      @Override
+      public AccessToken answer(InvocationOnMock invocation) throws Throwable {
+        synchronized (lock) {
+          lock.wait();
+        }
+        return new AccessToken("", new Date(expiration));
+      }
+    });
+
+    // Force a synchronous refresh.  This ought to wait until a refresh happening in another thread
+    // complets.
+    Callable<Void> syncRefreshCallable = new Callable<Void>() {
+      @Override
+      public Void call() throws Exception {
+        underTest.syncRefresh();
+        return null;
+      }
+    };
+
+    underTest = new RefreshingOAuth2CredentialsInterceptor(executorService, credentials);
+
+    // At this point, the access token wasn't retrieved yet. The
+    // RefreshingOAuth2CredentialsInterceptor considers null to be Expired.
+    Assert.assertEquals(CacheState.Expired,
+      RefreshingOAuth2CredentialsInterceptor.getCacheState(underTest.headerCache.get()));
+
+    Future<Void> future = executorService.submit(syncRefreshCallable);
+    
+    // let the other Threads have a turn so that 
+    Thread.yield();
+    synchronized(lock) {
+      lock.notifyAll();
+    }
+
+    // Try to get the access token, which should be calculated at this point.  There's
+    // a possibility that some hanging occurs in the test code, so timeout after 1 second if
+    // there is no response an.
+    future.get(1000, TimeUnit.MILLISECONDS);
```

Comment:
Why wait for 1 second here? either we get the result or TimeoutException.
In case of TimeoutException, the test fails.